packages/sacado/src/new_design/Sacado_Fad_Exp_Atomic.hpp

```diff
@@ -74,261 +74,307 @@ namespace Sacado {
-	  return_type return_val = *dest;
-	  *dest                  = op.apply(return_val, val);
-	  Kokkos::memory_fence();
-	  if (threadIdx.x == 0)
-	    Kokkos::Impl::unlock_address_cuda_space((void*)dest_val);
-	  return return_val;
-	}
-	else {
-	  return_type return_val;
-	  // This is a way to (hopefully) avoid dead lock in a warp
-	  int done                 = 0;
-	  unsigned int mask        =  __activemask() ;
-	  unsigned int active      = __ballot_sync(mask, 1);
-	  unsigned int done_active = 0;
-	  while (active != done_active) {
-	    if (!done) {
-	      if (Kokkos::Impl::lock_address_cuda_space((void*)dest_val)) {
-		Kokkos::memory_fence();
-		return_val = *dest;
-		*dest      = op.apply(return_val, val);
-		Kokkos::memory_fence();
-		Kokkos::Impl::unlock_address_cuda_space((void*)dest_val);
-		done = 1;
-	      }
-	    }
-	    done_active = __ballot_sync(mask, done);
-	  }
-	  return return_val;
-	}
-#elif defined(KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HIP_GPU)
-	// It is not allowed to define SACADO_VIEW_CUDA_HIERARCHICAL or
-	// SACADO_VIEW_CUDA_HIERARCHICAL_DFAD and use Sacado inside a team-based
-	// kernel without Sacado hierarchical parallelism.  So use the
-	// team-based version only if blockDim.x > 1 (i.e., a team policy)
+      }
+
+      // Helper function to decide if we are using team-based parallelism
+      SACADO_INLINE_FUNCTION
+      bool atomics_use_team() {
 #if defined(SACADO_VIEW_CUDA_HIERARCHICAL) || defined(SACADO_VIEW_CUDA_HIERARCHICAL_DFAD)
-	const bool use_team = (blockDim.x > 1);
+        // It is not allowed to define SACADO_VIEW_CUDA_HIERARCHICAL or
+        // SACADO_VIEW_CUDA_HIERARCHICAL_DFAD and use Sacado inside a team-based
+        // kernel without Sacado hierarchical parallelism.  So use the
+        // team-based version only if blockDim.x > 1 (i.e., a team policy)
+        return (blockDim.x > 1);
 #else
-	const bool use_team = false;
-#endif
-	if (use_team) {
-	  int go = 1;
-	  while (go) {
-	    if (threadIdx.x == 0)
-	      go = !Kokkos::Impl::lock_address_hip_space((void*)dest_val);
-	    go = Kokkos::Experimental::shfl(go, 0, blockDim.x);
-	  }
-	  Kokkos::memory_fence();
-	  return_type return_val = *dest;
-	  *dest                  = op.apply(return_val, val);
-	  Kokkos::memory_fence();
-	  if (threadIdx.x == 0)
-	    Kokkos::Impl::unlock_address_hip_space((void*)dest_val);
-	  return return_val;
-	}
-	else {
-	  return_type return_val;
-	  int done                 = 0;
-	  unsigned int active      = __ballot(1);
-	  unsigned int done_active = 0;
-	  while (active != done_active) {
-	    if (!done) {
-	      if (Kokkos::Impl::lock_address_hip_space((void*)dest_val)) {
-		return_val = *dest;
-		*dest      = op.apply(return_val, val);
-		Kokkos::Impl::unlock_address_hip_space((void*)dest_val);
-		done = 1;
-	      }
-	    }
-	    done_active = __ballot(done);
-	  }
-	  return return_val;
-	}
+        return false;
 #endif
       }
 
+#if defined(DESUL_HAVE_CUDA_ATOMICS)
```

Comment:
I'd use `KOKKOS_ENABLE_CUDA` instead (stronger guarantee that it won't change and other developers more familiar with the macro)

Reply:
Done.